### DISCUSSION TITLE
Descargar archivo de Facebook de la app de whatsapp de meta developers

### DISCUSSION THREAD

**Discussion by jeanp74** [community/community#130052] [Programming Help]
### Select Topic Area

Question

### Body

Hola, estoy realizando una integración con la app de Whatsapp de meta developers de Facebook.

Estoy realizando pruebas y al momento de enviar un mensaje al numero de whatsapp de la app me trae el id con el que yo puedo descargar ese archivo, al momento de hacerlo por la aplicación Postman me trae el archivo correctamente, pero luego al momento de querer replicarlo en php ya que así funciona el sitio web no me descarga el archivo correctamente como si estuviera dañado, lo estoy realizando con cURL y así tengo el código: 

`
    <?php
    date_default_timezone_set('America/Bogota');
    header("Cache-control: no-cache");

    // Tu token de acceso
    $token = 'Token proporcionado por Faceboock';

    // ID del archivo subido (id_documento)
    $id_documento = 'id del documento';

    // URL para obtener la información del archivo
    $url = "https://graph.facebook.com/v19.0/$id_documento";

    // Configurar los encabezados de la solicitud
    $headers = [
        "Authorization: Bearer $token",
        "Content-Type: application/json"
    ];

    // Inicializar cURL para obtener la URL de descarga y tipo MIME
    $curl = curl_init();
    curl_setopt($curl, CURLOPT_URL, $url);
    curl_setopt($curl, CURLOPT_HTTPHEADER, $headers);
    curl_setopt($curl, CURLOPT_RETURNTRANSFER, true);

    // Ejecutar la solicitud
    $respuesta = curl_exec($curl);

    // Verificar si hay errores
    if ($respuesta === false) {
        $error = curl_error($curl);
        echo "Error: $error";
        die("ERROR AL ENVIAR LA URL");
    }

    // Cerrar cURL inicial
    curl_close($curl);

    $data = json_decode($respuesta, true);

    if ($data != '' && $data != NULL) {
        $url_descarga = $data['url'];
        $tipo_mime = $data['mime_type'];

        // Mapa MIME a extensión
        $mime_map = [
            'application/pdf' => 'pdf',
            'image/jpg' => 'jpg',
            'image/jpeg' => 'jpeg',
            'image/png' => 'png',
            'video/mp4' => 'mp4',
        ];

        $extension = isset($mime_map[$tipo_mime]) ? $mime_map[$tipo_mime] : 'bin';

        // Configurar los encabezados de la solicitud
        $headers = [
            "Authorization: Bearer $token"
        ];

        // Inicializar cURL para descargar el archivo
        $curl_archivo = curl_init();
        curl_setopt($curl_archivo, CURLOPT_URL, $url_descarga);
        curl_setopt($curl_archivo, CURLOPT_HTTPHEADER, $headers); // Usar el mismo header con el token
        curl_setopt($curl_archivo, CURLOPT_RETURNTRANSFER, true);
        curl_setopt($curl_archivo, CURLOPT_HEADER, 0);
        curl_setopt($curl_archivo, CURLOPT_FOLLOWLOCATION, true);
        curl_setopt($curl_archivo, CURLOPT_SSL_VERIFYPEER, false); // Desactivar verificación SSL si es necesario
        curl_setopt($curl_archivo, CURLOPT_AUTOREFERER, true);

        // Ejecutar la solicitud de descarga
        $datos_archivo = curl_exec($curl_archivo);

        // Verificar si hay errores en la descarga
        if ($datos_archivo === false) {
            $error = curl_error($curl_archivo);
            die("Error de descarga: $error");
        }

        // Obtener el código de respuesta HTTP
        $httpCode = curl_getinfo($curl_archivo, CURLINFO_HTTP_CODE);

        // Verificar si la descarga fue exitosa (código 200)
        if ($httpCode !== 200) {
            die('Error: No se pudo descargar el archivo. HTTP Code: ' . $httpCode);
        }

        // Cerrar cURL de descarga
        curl_close($curl_archivo);
        
        // Guardar el archivo en el servidor con la extensión dinámica
        $archivo = "archivo_descargado.$extension";

        // Configurar las cabeceras HTTP para la descarga
        header("Content-Description: File Transfer");
        header("Content-Type: $tipo_mime");
        header("Content-Disposition: attachment; filename=$archivo");
        header("Content-Transfer-Encoding: binary");
        header("Expires: 0");
        header("Cache-Control: must-revalidate");
        header("Pragma: public");
        header("Content-Length: " . strlen($datos_archivo));

        // Enviar el contenido del archivo
        echo $datos_archivo;
    }
    ?>
`

Cuando descarga el archivo me arroja un error que dice: 

"Error
No se pudo cargar el documento PDF."

Pero el archivo si tiene el peso correcto, y también al momento de imprimir la respuesta del cURL me arroja lo siguiente: 
![Captura de pantalla 2024-06-27 124920](https://github.com/community/community/assets/164540906/7e543e52-aead-481a-b3bf-ce534348340d)
 
Tengo el navegador actualizado y todo y sigue sin funcionar.

Sinceramente llevo un año programando y soy nuevo en esto de cURL y no entiendo el por que me falla.

**Comment by ebndev** [comment]
Hi @jeanp74, thanks for participating!

Unfortunately, we don’t currently have moderators for languages other than English. Until that changes, we need to ask that everyone use English here in the GitHub Community when posting to ensure discussions are adhering to our [Code of Conduct](https://github.com/community/community/blob/920a928ffe3ae4863b64a4ea26917046e11f5c1a/CODE_OF_CONDUCT.md). 

We’ll be locking this post for now, but please feel encouraged to create a new discussion in English if you still need support.
